File: src/DynamicImages/Client/src/models/property-path.ts
/**
 * A binding's `propertyAlias` may be a dotted path - `author.mainImage` - which follows the
 * content reference in its first segment and reads the last segment on the node it lands on.
 *
 * This is the client's half of `Core/Rendering/PropertyPath.cs`, kept deliberately small: it only
 * has to split a stored alias into the two dropdowns the inspector shows and put it back together.
 * The server is what actually resolves a path.
 */

/** How many content references a path may follow. Mirrors `PropertyPath.MaxHops`. */
export const MAX_HOPS = 3;

/** Whether an alias names a property on a linked node rather than on this one. */
export const isPath = (alias: string | null | undefined): boolean => hopCount(alias) > 0;

/** How many references an alias follows. */
export const hopCount = (alias: string | null | undefined): number => Math.max(0, segmentsOf(alias).length - 1);

/**
 * Splits a stored alias into the root property and everything after it.
 *
 * The whole remainder stays in the tail - `splitPath("a.b.c")` gives `{ root: "a", tail: "b.c" }`.
 * That is the correct decode for a two-dropdown UI: a three-segment path typed by hand or imported
 * from JSON then survives a round trip through the inspector untouched, rather than being
 * flattened to two segments the moment the editor opens the layer.
 */
export function splitPath(alias: string | null | undefined): { root: string; tail: string } {
  const segments = segmentsOf(alias);

  return { root: segments[0] ?? "", tail: segments.slice(1).join(".") };
}

/** Puts a root and a tail back together. An empty tail gives the bare root. */
export function joinPath(root: string, tail: string | null | undefined): string {
  const trimmedRoot = (root ?? "").trim();
  const trimmedTail = (tail ?? "").trim();

  if (!trimmedRoot) return "";

  return trimmedTail ? `${trimmedRoot}.${trimmedTail}` : trimmedRoot;
}

/** Empty segments are dropped, so `"author."` degrades to the bare alias - defensive, not a feature. */
const segmentsOf = (alias: string | null | undefined): string[] =>
  (alias ?? "")
    .split(".")
    .map((segment) => segment.trim())
    .filter((segment) => segment.length > 0);

File: src/DynamicImages/Core/Rendering/PropertyPath.cs
namespace Umbraco.Community.DynamicImages.Core.Rendering;

/// <summary>
/// A dotted property alias - <c>author.mainImage</c> - split into the hops that follow a content
/// reference and the alias that is finally read.
/// <para>
/// The dotted string rides in the existing <c>propertyAlias</c> fields, so uSync serialisation,
/// the template JSON and <c>{prop:author.jobTitle}</c> expressions all carry it unchanged.
/// </para>
/// </summary>
/// <param name="First">The first segment - the property read on the node being rendered.</param>
/// <param name="Hops">Every segment except the last: the references that are followed.</param>
/// <param name="Last">The alias read on the node the hops land on. Equals <paramref name="First"/> for a bare alias.</param>
/// <param name="IsTooDeep">The path follows more references than <see cref="MaxHops"/> allows.</param>
public readonly record struct PropertyPath(string First, IReadOnlyList<string> Hops, string Last, bool IsTooDeep)
{
    /// <summary>
    /// How many content references a path may follow. An over-cap path resolves to nothing rather
    /// than being truncated: truncating would quietly read the wrong property, and the validator's
    /// <c>PropertyPathTooDeep</c> warning is how the editor finds out instead.
    /// </summary>
    public const int MaxHops = 3;

    /// <summary>The number of segments: one more than the number of hops.</summary>
    public int SegmentCount => Hops.Count + 1;

    /// <summary>Whether the alias names a property on a linked node rather than on this one.</summary>
    public static bool IsPath(string? alias) => Parse(alias).Hops.Count > 0;

    /// <summary>
    /// Splits a dotted alias. Empty segments are dropped, so <c>"author."</c> degrades to the bare
    /// alias - defensive, not a feature.
    /// </summary>
    public static PropertyPath Parse(string? alias)
    {
        if (string.IsNullOrWhiteSpace(alias)) return new PropertyPath(string.Empty, [], string.Empty, false);

        var segments = alias
            .Split('.', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .ToArray();

        if (segments.Length == 0) return new PropertyPath(string.Empty, [], string.Empty, false);

        var hops = segments[..^1];

        return new PropertyPath(segments[0], hops, segments[^1], hops.Length > MaxHops);
    }
}
